rfc-134-govuk-wide-session-cookie-and-login.md

```diff
@@ -0,0 +1,354 @@
+# A GOV.UK-wide session cookie & login
+
+## Summary
+
+We propose introducing a new cookie, `__Host-govuk_account_session`,
+which will be an essential cookie, but only set when a user signs in
+to use personalised parts of GOV.UK (currently just the Transition
+Checker).
+
+Similarly to how our A/B tests work we will manage this cookie at the
+Fastly layer, in Varnish Configuration Language (VCL), and use custom
+request headers to pass the cookie value to our apps. We will also use
+custom response headers to set a new cookie value.
+
+We will create a new app to provide the internal account-supporting
+API, and extend frontend to handle the login and logout process.
+
+## Problem
+
+The GOV.UK Account team have launched an experiment on the Transition
+Checker, allowing users to sign up to save their
+results. [finder-frontend sets a session cookie][ff-session]—an
+encrypted cookie—containing the user ID and OAuth tokens used to
+update the data we hold on them.
+
+This works fine for one app, but there are problems with this approach
+when we try to scale to more than one app.
+
+[ff-session]: https://github.com/alphagov/finder-frontend/blob/5b6101e8eb027028a102ab96bcef6ff752b9c558/app/controllers/sessions_controller.rb#L23-L28
+
+### We need to pass cookies to our apps
+
+Our Nginx configuration blocks most cookies, [which we disabled for
+the Transition Checker][tc-cookie-pr]. If multiple apps need to
+consume the session cookie, then this blocking will be disabled for
+ever-increasing chunks of GOV.UK.
+
+[tc-cookie-pr]: https://github.com/alphagov/govuk-puppet/pull/10788
+
+### Apps need to share the same session cookie
```

Comment:
This isn't true in the general case, which is that services will have their own cookies, and will redirect users in an OIDC flow to the central authn service if the user isn't logged in, but needs to be. There will need to be some common standards to make that work - including a way for users to log themselves out, and some agreement about cookie lifetimes.

Not that GOV.UK _has_ to have each of is apps implement authentication against the central service separately: where you draw the lines between your services is up to you. But the architecture must allow for operationally independent services operated by different departments in any event, and the more GOV.UK can itself behave like other services will, the more good patterns can be embedded from the start, and the more the OIDC flows will be properly dog-fooded. And the fewer surprises there will be when we work with other government departments to make their services feel properly integrated with the GOV.UK Account.

Reply:
Right, this RFC is specifically about `www.gov.uk`, not everything under the `gov.uk` domain.

Reply:
I get that, I still think that treating different services that the GOV.UK team happen to be running more separately might have some merit, and if you're excluding _in general_ then it's worth stating why. Seeing as there are some benefits which I mentioned :-)

Reply:
Ok, that makes sense.  I think the main problems with apps on the www.gov.uk domain authenticating separately are:

- It's probably confusing to users if they move between different parts of GOV.UK and their logged-in state changes.  We've already found people being confused at the difference between being logged in to the Transition Checker on www.gov.uk and the OIDC thing on www.account.publishing.service.gov.uk, and they're not even on the same domain.
- It adds friction to personalisation, where we'd have to get users to re-authenticate to see personalised content served by different apps.
- It makes it harder to show some global logged-in state, eg a "logout" link in the header, because the user is individually logged into separate parts of GOV.UK.

But for services *not* on www.gov.uk, eg if there's some new big thing and the GOV.UK team spins up a prepare-for-meteor-impact.service.gov.uk, *that* should authenticate separately, because it's not on the same domain.

My thinking is that the microservices which make up GOV.UK (or prepare-for-meteor-impact.service.gov.uk, or whatever) are an implementation detail which shouldn't be exposed to the user, so you'd ideally want all the microservices that make up one service to share their authenticated-ness.

Reply:
I think it's confusing for users if logged-in state isn't maintained between _all_ services. This will have to be tackled at some point (although not necessarily for this RFC of course!).

The whole point of gov.uk is that users see a single thing and that the question of which department or team provides _any_ service that central government is responsible for is an implementation detail.

Reply:
> treating different services that the GOV.UK team happen to be running more separately might have some merit
I was thinking that this approach isn't practical as there are 60+ services which make up GOV.UK Publishing, will the GOV.UK Account want to support that many connections for _one_ client. Then scaling that to each of the micro services which make up the 300+ services across Government.

There's also an OpenID Connect spec for this Open ID [Connect Federation](https://openid.net/specs/openid-connect-federation-1_0.html), which I need to read and see how compatible it is with this RFC.

I would like to think that the GOV.UK Account will consider GOV.UK Publishing as a single service and allow it to manage users sessions in a way is appropriate for it's architecture. An implementation detail as @barrucadu says.

Reply:
Also, if we do end up wanting to split GOV.UK Publishing into separate clients, that can always happen later.